core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -100,36 +116,48 @@ public Schema union(Schema union, List<Schema> options) {
   }
 
   @Override
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
   public Schema array(Schema array, Schema element) {
-    if (array.getLogicalType() instanceof LogicalMap) {
+    if (array.getLogicalType() instanceof LogicalMap || AvroSchemaUtil.isKeyValueSchema(array.getElementType())) {
       Schema keyValue = array.getElementType();
-      int keyId = AvroSchemaUtil.getFieldId(keyValue.getField("key"));
-      int valueId = AvroSchemaUtil.getFieldId(keyValue.getField("value"));
+      Integer keyId = fieldId(keyValue.getField("key"));
+      Integer valueId = fieldId(keyValue.getField("value"));
+      if (keyId == null) {
+        if (valueId != null) {
+          LOG.warn("Map schema {} has value id but not key id", array);
+        }
+        return null;
+      }
 
       // if either key or value is selected, the whole map must be projected
       if (selectedIds.contains(keyId) || selectedIds.contains(valueId)) {
-        return array;
+        return complexMapWithIds(array, keyId, valueId);
       } else if (element != null) {
-        if (keyValue.getField("value").schema() != element.getField("value").schema()) {
+        if (keyValue.getField("key").schema() != element.getField("key").schema() ||
+            keyValue.getField("value").schema() != element.getField("value").schema()) {
```

Comment:
Because the key is always fully projected, this doesn't need to check the key for changes.

Reply:
key schema could be different if new field ids are assigned to fields under the key schema

Reply:
Could we do this check separately? e.g make sure that key schemas are not projected (but they could contain new ids)

Reply:
Yes, we will need to make sure the key schema is structurally identical. You can use `SchemaNormalization.parsingFingerprint64` for it.

Reply:
Thanks I'll try this out